File: google-maps/android/src/main/java/com/capacitorjs/plugins/googlemaps/GoogleMapCameraConfig.kt
package com.capacitorjs.plugins.googlemaps

import com.google.android.gms.maps.model.LatLng
import org.json.JSONObject

class GoogleMapCameraConfig(fromJSONObject: JSONObject) {
    var coordinate: LatLng? = null
    var zoom: Double? = null
    var angle: Double? = null
    var bearing: Double? = null
    var animate: Boolean? = null
    var animationDuration: Double? = null

    init {
        if (fromJSONObject.has("zoom")) {
            zoom = fromJSONObject.getDouble("zoom")
        }

        if(fromJSONObject.has("angle")) {
            angle = fromJSONObject.getDouble("angle")
        }

        if (fromJSONObject.has("bearing")) {
            bearing = fromJSONObject.getDouble("bearing")
        }

        if (fromJSONObject.has("animate")) {
            animate = fromJSONObject.getBoolean("animate")
        }

        if (fromJSONObject.has("animationDuration")) {
            animationDuration = fromJSONObject.getDouble("animationDuration")
        }

        if (fromJSONObject.has("coordinate")) {
            val coordinateJSONObject = fromJSONObject.getJSONObject("coordinate")
            if(!coordinateJSONObject.has("lat") || !coordinateJSONObject.has("lng")) {
                throw InvalidArgumentsError("LatLng object is missing the required 'lat' and/or 'lng' property")
            }

            val lat = coordinateJSONObject.getDouble("lat")
            val lng = coordinateJSONObject.getDouble("lng")
            coordinate = LatLng(lat, lng)
        } else {
            coordinate = null
        }
    }


}

File: google-maps/ios/Plugin/GoogleMapCameraConfig.swift
import Foundation
import Capacitor

public struct GoogleMapCameraConfig {
    let coordinate: LatLng?
    let zoom: Float?
    let bearing: Double?
    let angle: Double?
    let animate: Bool?
    let animationDuration: Double?

    init(fromJSObject: JSObject) throws {
        zoom = fromJSObject["zoom"] as? Float
        bearing = fromJSObject["bearing"] as? Double
        angle = fromJSObject["angle"] as? Double
        animate = fromJSObject["animate"] as? Bool
        animationDuration = fromJSObject["animationDuration"] as? Double

        if let latLngObj = fromJSObject["coordinate"] as? JSObject {
            guard let lat = latLngObj["lat"] as? Double, let lng = latLngObj["lng"] as? Double else {
                throw GoogleMapErrors.invalidArguments("LatLng object is missing the required 'lat' and/or 'lng' property")
            }

            self.coordinate = LatLng(lat: lat, lng: lng)
        } else {
            self.coordinate = nil
        }
    }
}
